pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -267,10 +270,30 @@ func NewController(kubeClient kubelib.Client, options Options) *Controller {
 		})
 	})
 	registerHandlers(c.pods.informer, c.queue, "Pods", c.pods.onEvent, nil)
-
+	if err := updateSPIFFERegistry(kubeClient, c.clusterID); err != nil {
+		log.Errorf("Failed to update the trust domain for cluster %v: %v", c.clusterID, err)
+	}
 	return c
 }
 
+func updateSPIFFERegistry(kubeClient kubelib.Client, clusterID string) error {
+	cm, err := kubeClient.CoreV1().ConfigMaps(IstioNamespace).Get(context.TODO(), constants.IstioMeshConfigConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("unable to read config map for trust domain: %v", err)
+	}
+	if cm == nil || cm.Data == nil {
+		log.Debugf("MeshConfig empty, skip trust domain updates")
+		return nil
+	}
+	yml := cm.Data[constants.IstioMeshConfigEntryName]
+	cfg, err := mesh.ApplyMeshConfigDefaults(yml)
+	if err != nil {
+		return fmt.Errorf("unable to deserialize mesh config from YAML %v, error: %v", yml, err)
+	}
+	spiffe.SetTrustDomainByCluster(clusterID, cfg.TrustDomain)
```

Comment:
+1, the spifeeSet/Get trust domain should be removed.

Reply:
Yes this is not good. The side effects will make things very hard to reason about. To make the entire controller module more reusable, lets turn the spiffe ting into an object? I understand that this will permeate throughout the code base but better late than never.

Reply:
Singleton has many flaws like different parts of code can set/get the shared resource without dependency explicitly declared. @rshriram is that the paint point you mentioned? Just want to see if I understand the point completely.

I can make this as a SPIFFE object, consuming information from MeshConfig,

```go
type SpiffeRegistry struct {
  // func GetTrustDomain(clusterID string)
}
```

This struct would then have to be initiated in the xDS server and pass all the way down to all the places it's needed to replace the Singleton's GetXX method, requiring maybe lots of constructor parameter change.

@howardjohn would that address the concern?

Reply:
addressed, since reading from meshconfig now.